### DISCUSSION TITLE
Concurrency for FastAPI application using in-memory DuckDB

### DISCUSSION THREAD

**Discussion by jfvalenzuelas** [duckdb/duckdb#13719] [Q&A]
Hi,

I'm developing an API using FastAPI and DuckDB and while doing a stress test with Locust, I had the issue that the application process stoped responding. There is no error nor message on the terminal.

Something important to tell here is that we are running the application with only one worker. This is due to the use of a in-memory caching that I'm doing and that I need to refresh after certain operations. To avoid to coordinate this refresh at the same time for all workers, if using multiple, this is running only on one.

This is a minimal dummy example of what it is implemented and should be reproducible when using Locust.

``` python
from faker import Faker
from faker.providers import internet
import duckdb
import pandas as pd
from fastapi import FastAPI

app = FastAPI()

fake = Faker()
fake.add_provider(internet)

conn = duckdb.connect(database = ":memory:")

def fake_row() -> list:
    return [fake.name(), fake.address(), fake.ascii_email(), fake.ipv4_private()]

data = [fake_row() for _ in range(100)]
df = pd.DataFrame(data, columns=['Name', 'Address', 'Email', 'IPV4'])

conn.register("my_table", df)

@app.get("/my-endpoint")
def my_endpoint():
    res = conn.sql("SELECT * FROM my_table")
    tmp_df = pd.DataFrame(res.fetchall(), columns=[col[0] for col in res.description])
    print(tmp_df)
    return True
```

This is the locustfile.py:

``` python
import os
import base64
from locust import HttpUser, task, between


class QuickstartUser(HttpUser):
    wait_time = between(1, 5) 

    @task
    def test(self):
        url = "/my-endpoint"

        response = self.client.get(
            url
        )

        assert response.status_code == 200
```

These are the versions that I have installed in my environment:

``` text
duckdb==1.0.0
Faker==28.1.0
fastapi==0.112.2
locust==2.31.5
uvicorn==0.30.6
```

I also noticed that if we use `with conn.cursor() as cursor: ...` the cursor cannot see (has no access) to the view `my_table` registered with the conn.

did somehow run into this already or does someone have any advices or tips?

**UPDATE**

Instead of registering the dataframe as a view, I have created a table based out of it, and now all new connections via a cursor seem to have access to this created table:

``` python
from faker import Faker
from faker.providers import internet
import duckdb
import pandas as pd
from fastapi import FastAPI

app = FastAPI()

fake = Faker()
fake.add_provider(internet)

conn = duckdb.connect(database = ":memory:")

def fake_row() -> list:
    return [fake.name(), fake.address(), fake.ascii_email(), fake.ipv4_private()]

data = [fake_row() for _ in range(100)]
df = pd.DataFrame(data, columns=['Name', 'Address', 'Email', 'IPV4'])

conn.sql("CREATE TABLE my_table AS SELECT * FROM df")
conn.sql("INSERT INTO my_table SELECT * FROM df")

@app.get("/my-endpoint")
def my_endpoint():
    with conn.cursor() as cursor:
        res = cursor.execute("SELECT * FROM my_table")
        tmp_df = pd.DataFrame(res.fetchall(), columns=[col[0] for col in res.description])
        print(tmp_df)
        return True
```

Can somehow explain why a view is not available for all connections but only for the connection that registered the view?

Thank you all!

**Comment by robotslacker** [comment]
You can try 
`conn = duckdb.connect(database = ":memory:xxx")`

Here, xxx is a memory segment name.  keep all your memory connections use same name.

**Reply by jfvalenzuelas** [reply]
Thanks for the reply!

I tried this but it does not work when you use duckdb.register().

I noticed that the "view" that is registered using this function is only accessible or visible for the connection that was used for the execution. New connections (also with cursor) cannot access it.

I have changed the code to create or replace a view using duckdb.sql() and this way, the view is also accessible for every other connection opened using cursors.
